src/Microsoft.Azure.Functions.Extensions.Mcp/Trigger/ToolReturnValueBinder.cs

```diff
@@ -0,0 +1,105 @@
+
+using System.Collections.ObjectModel;
+using System.Text.Json;
+using Microsoft.Azure.Functions.Extensions.Mcp.Serialization;
+using Microsoft.Azure.WebJobs.Host.Bindings;
+using ModelContextProtocol;
+using ModelContextProtocol.Protocol;
+
+namespace Microsoft.Azure.Functions.Extensions.Mcp;
+
+internal sealed class ToolReturnValueBinder(CallToolExecutionContext executionContext) : IValueBinder
+{
+    public Type Type { get; } = typeof(object);
+
+    public Task SetValueAsync(object value, CancellationToken cancellationToken)
+    {
+        if (value is null)
+        {
+            executionContext.SetResult(null!);
+            return Task.CompletedTask;
+        }
+
+        if (value is not string jsonString)
+        {
+            throw new ArgumentException("Expected JSON string.", nameof(value));
+        }
+
+        var result = JsonSerializer.Deserialize<McpToolResult>(jsonString, McpJsonSerializerOptions.DefaultOptions)
+                     ?? throw new InvalidOperationException("The function return value could not be deserialized to a valid McpToolResult.");
+
+
+        if (string.IsNullOrEmpty(result.Type))
+        {
+            throw new InvalidOperationException($"The McpToolResult '{nameof(result.Type)}' cannot be null or empty.");
+        }
+
+        try
+        {
+            Collection<ContentBlock> contentBlocks = DeserializeToContentBlockCollection(result);
+
+            if (contentBlocks.Count == 0)
+            {
+                throw new InvalidOperationException("No content items were produced from the McpToolResult.");
+            }
+
+            executionContext.SetResult(new CallToolResult
+            {
+                Content = contentBlocks
+            });
+        }
+        catch (JsonException ex)
+        {
+            throw new InvalidOperationException($"Failed to deserialize content for type '{result.Type}'.", ex);
+        }
+
+        return Task.CompletedTask;
+    }
+
+    public Task<object> GetValueAsync() => throw new NotSupportedException();
+
+    public string ToInvokeString() => string.Empty;
+
+    private static Collection<ContentBlock> DeserializeToContentBlockCollection(McpToolResult result)
+    {
+        var blocks = new Collection<ContentBlock>();
+
+        // Explicit multi-content contract
+        if (string.Equals(result.Type, McpConstants.ToolResultContentTypes.MultiContentResult, StringComparison.OrdinalIgnoreCase))
+        {
+
+            var collection = JsonSerializer.Deserialize<Collection<ContentBlock>>(result.Content!, McpJsonSerializerOptions.DefaultOptions)
+                ?? throw new InvalidOperationException("Failed to deserialize multi-content result.");
+
+            foreach (var item in collection)
+            {
+                blocks.Add(item);
+            }
+        }
+        else
+        {
+            // Otherwise, handle single content block
+            var contentBlock = result.Type switch
```

Comment:
`ContentBlock` provides a JSON converter that will handle conversion to the appropriate types based on the `Type` value.

The following should be sufficient, and will avoid the need to switch on the type or have the `DeserializeContentBlock` method:

``` csharp
var contentBlock = JsonSerializer.Deserialize<ContentBlock>(result.Content!, McpJsonUtilities.DefaultOptions)
    ?? throw new InvalidOperationException("Failed to deserialize content block."
```